trie/nodeset.go

```diff
@@ -79,81 +86,82 @@ func (n *nodeWithPrev) unwrap() *memoryNode {
 // memorySize returns the total memory size used by this node. It overloads
 // the function in memoryNode by counting the size of previous value as well.
 // nolint: unused
-func (n *nodeWithPrev) memorySize(key int) int {
-	return n.memoryNode.memorySize(key) + len(n.prev)
-}
-
-// nodesWithOrder represents a collection of dirty nodes which includes
-// newly-inserted and updated nodes. The modification order of all nodes
-// is represented by order list.
-type nodesWithOrder struct {
-	order []string                 // the path list of dirty nodes, sort by insertion order
-	nodes map[string]*nodeWithPrev // the map of dirty nodes, keyed by node path
+func (n *nodeWithPrev) memorySize(pathlen int) int {
+	return n.memoryNode.memorySize(pathlen) + len(n.prev)
 }
 
 // NodeSet contains all dirty nodes collected during the commit operation.
 // Each node is keyed by path. It's not thread-safe to use.
 type NodeSet struct {
-	owner   common.Hash       // the identifier of the trie
-	updates *nodesWithOrder   // the set of updated nodes(newly inserted, updated)
-	deletes map[string][]byte // the map of deleted nodes, keyed by node
-	leaves  []*leaf           // the list of dirty leaves
+	owner  common.Hash            // the identifier of the trie
+	nodes  map[string]*memoryNode // the set of dirty nodes(inserted, updated, deleted)
+	leaves []*leaf                // the list of dirty leaves
+
+	// The list of accessed nodes, which records the original node value.
+	// The origin value is expected to be nil for newly inserted node
+	// and is expected to be non-nil for other types(updated, deleted).
+	accessList map[string][]byte
 }
 
 // NewNodeSet initializes an empty node set to be used for tracking dirty nodes
 // from a specific account or storage trie. The owner is zero for the account
 // trie and the owning account address hash for storage tries.
-func NewNodeSet(owner common.Hash) *NodeSet {
+func NewNodeSet(owner common.Hash, accessList map[string][]byte) *NodeSet {
 	return &NodeSet{
-		owner: owner,
-		updates: &nodesWithOrder{
-			nodes: make(map[string]*nodeWithPrev),
-		},
-		deletes: make(map[string][]byte),
+		owner:      owner,
+		nodes:      make(map[string]*memoryNode),
+		accessList: accessList,
 	}
 }
 
-/*
-// NewNodeSetWithDeletion initializes the nodeset with provided deletion set.
-func NewNodeSetWithDeletion(owner common.Hash, paths [][]byte, prev [][]byte) *NodeSet {
-	set := NewNodeSet(owner)
-	for i, path := range paths {
-		set.markDeleted(path, prev[i])
+// forEachWithOrder iterates the dirty nodes with the specified order.
+// If topToBottom is true:
+//
+//	then the order of iteration is top to bottom, left to right.
+//
+// If topToBottom is false:
+//
+//	then the order of iteration is bottom to top, right to left.
+func (set *NodeSet) forEachWithOrder(topToBottom bool, callback func(path string, n *memoryNode)) {
```

Comment:
Remove this topToBottom option.

Reply:
Here's a more comprehensive suggestion: 


```diff
diff --git a/trie/database.go b/trie/database.go
index 442f594c16..c5d94ef137 100644
--- a/trie/database.go
+++ b/trie/database.go
@@ -792,10 +792,7 @@ func (db *Database) Update(nodes *MergedNodeSet) error {
 	}
 	for _, owner := range order {
 		subset := nodes.sets[owner]
-		subset.forEachWithOrder(false, func(path string, n *memoryNode) {
-			if n.isDeleted() {
-				return // ignore deletion
-			}
+		subset.forEachExisting(func(path string, n *memoryNode) {
 			db.insert(n.hash, int(n.size), n.node)
 		})
 	}
diff --git a/trie/nodeset.go b/trie/nodeset.go
index 1908eb18af..4252d8a591 100644
--- a/trie/nodeset.go
+++ b/trie/nodeset.go
@@ -114,24 +114,18 @@ func NewNodeSet(owner common.Hash, accessList map[string][]byte) *NodeSet {
 	}
 }
 
-// forEachWithOrder iterates the dirty nodes with the specified order.
-// If topToBottom is true:
-//
-//	then the order of iteration is top to bottom, left to right.
-//
-// If topToBottom is false:
-//
-//	then the order of iteration is bottom to top, right to left.
-func (set *NodeSet) forEachWithOrder(topToBottom bool, callback func(path string, n *memoryNode)) {
+// forEachExisting iterates the dirty nodes.
+// The order of iteration is bottom to top, right to left.
+// The callback will _not_ be invoked for deleted nodes.
+func (set *NodeSet) forEachExisting(callback func(path string, n *memoryNode)) {
 	var paths sort.StringSlice
-	for path := range set.nodes {
-		paths = append(paths, path)
-	}
-	if topToBottom {
-		paths.Sort()
-	} else {
-		sort.Sort(sort.Reverse(paths))
+	for path, node := range set.nodes {
+		if !node.isDeleted() {
+			paths = append(paths, path)
+		}
 	}
+	// Bottom-up, longest paths first
+	sort.Sort(sort.Reverse(paths))
 	for _, path := range paths {
 		callback(path, set.nodes[path])
 	}
```

Reply:
I still prefer to iterate deleted nodes as well. This function `forEachWithOrder` is used to iterate all nodes inside, no matter it's a deletion or update. It's just very special that we can't do deletion in hash-based scheme.

Although this function is only used by trie.Database, but filter out deleted nodes feels a bit un-general to me.

Reply:
Gotcha